source/extensions/tracers/zipkin/span_buffer.cc

```diff
@@ -16,22 +33,200 @@ bool SpanBuffer::addSpan(const Span& span) {
   return true;
 }
 
-std::string SpanBuffer::toStringifiedJsonArray() {
+SerializerPtr SpanBuffer::makeSerializer(
+    const envoy::config::trace::v2::ZipkinConfig::CollectorEndpointVersion& version,
+    const bool shared_span_context) {
+  switch (version) {
+  case envoy::config::trace::v2::ZipkinConfig::HTTP_JSON_V1:
+    return std::make_unique<JsonV1Serializer>();
+  case envoy::config::trace::v2::ZipkinConfig::HTTP_JSON:
+    return std::make_unique<JsonV2Serializer>(shared_span_context);
+  case envoy::config::trace::v2::ZipkinConfig::HTTP_PROTO:
+    return std::make_unique<ProtobufSerializer>(shared_span_context);
+  default:
+    // TODO(dio): Throw if it is required to be explicit when specifying version.
```

Comment:
I would default to HTTP_JSON v2 as that is the idiomatic default for Zipkin now and next to Zipkin core ecosystem most if not all third party custom collectors use V2 as well. V2 is a performance and wire size improvement over v1.

Reply:
Yeah. This is a mistake here though (I'll remove this line in the next commit). In the meantime, users are allowed to **SKIP** specifying `collector_endpoint_version` (which is not mandatory for now). It will default to `NOT_SET`. And then the current code (https://github.com/envoyproxy/envoy/pull/6985/files#diff-9d8d9d059ecfb58df313d5b1a3d9834eR150) set the default to `envoy::config::trace::v2::ZipkinConfig::HTTP_JSON_V1` for now because that is the current behavior. When we deprecate or remove `envoy::config::trace::v2::ZipkinConfig::HTTP_JSON_V1` and `envoy::config::trace::v2::ZipkinConfig::NOT_SET`, then we can force users to specify `collector_endpoint_version` via configuration.